File: UI/src/lib/types/dev-machine.ts
export type DevMachineStatus =
	| 'configuring'
	| 'queued'
	| 'spawning'
	| 'running'
	| 'paused'
	| 'stopping'
	| 'stopped'
	| 'tearing_down'
	| 'destroyed'
	| 'failed'
	| 'expired';

type DevMachineLaunchStatus = 'ready' | 'resuming' | 'pending';

type DevMachineEnvironmentStatus = 'pending' | 'building' | 'ready' | 'failed' | 'delete_requested';

interface DevMachineOperation {
	id: string;
	action: string;
	status: 'pending' | 'leased' | 'completed' | 'failed' | 'cancelled' | string;
	generation: number;
	idempotency_key: string;
	attempts: number;
	error_code?: string;
	error_message?: string;
	created_at: string;
	completed_at?: string;
}

type AgentRunStatus =
	| 'queued'
	| 'starting'
	| 'running'
	| 'waiting_input'
	| 'succeeded'
	| 'failed'
	| 'cancelled'
	| 'timeout';

export interface DevMachine {
	id: string;
	workspace_id: string;
	project_id?: string;
	issue_id?: string;
	created_by_user_id?: string;
	routing_key: string;
	name: string;
	status: DevMachineStatus;
	desired_status: DevMachineStatus;
	generation: number;
	repo_url?: string;
	repo_provider?: 'github';
	repo_owner?: string;
	repo_name?: string;
	base_branch?: string;
	working_branch?: string;
	machine_size: 'small' | 'medium' | 'large';
	cpu_millis: number;
	memory_mb: number;
	disk_gb: number;
	pids_limit: number;
	max_runtime_minutes: number;
	environment_id?: string;
	repository_affinity_id?: string;
	keep_running: boolean;
	environment_builder: boolean;
	last_activity_at?: string;
	delete_requested_at?: string;
	docker_network_name?: string;
	workspace_volume_name?: string;
	last_error_code?: string;
	last_error_message?: string;
	created_at: string;
	updated_at: string;
	started_at?: string;
	stopped_at?: string;
	expires_at: string;
	teardown_at?: string;
	destroyed_at?: string;
}

export interface DevMachineService {
	id: string;
	machine_id: string;
	agent_run_id?: string;
	service_type: 'ide' | 'terminal' | 'agent' | 'browser' | 'collector' | 'egress';
	service_key: string;
	container_id?: string;
	container_name: string;
	image_ref: string;
	status: string;
	health_status: string;
	health_message?: string;
	started_at?: string;
}

export interface DevMachineEvent {
	id: number;
	machine_id: string;
	agent_run_id?: string;
	source: string;
	event_type: string;
	payload: Record<string, unknown>;
	occurred_at: string;
}

export interface DevMachineLogChunk {
	id: number;
	agent_run_id?: string;
	service_id?: string;
	stream: 'stdout' | 'stderr' | 'pty' | 'system';
	sequence: number;
	content: string;
	truncated: boolean;
	created_at: string;
}

interface AgentArtifact {
	type: string;
	name: string;
	path: string;
	content_type?: string;
}

interface AgentRunResult {
	status: 'succeeded' | 'failed' | 'cancelled' | 'timeout';
	summary: string;
	changed_files: string[];
	commits: string[];
	branch?: string;
	pull_request_url?: string;
	tests_run: string[];
	test_status: 'passed' | 'failed' | 'not_run';
	risk_notes: string[];
	artifacts: AgentArtifact[];
}

export interface AgentRun {
	id: string;
	machine_id: string;
	issue_id?: string;
	checkout_id?: string;
	provider_id: string;
	mode: 'interactive' | 'autonomous';
	status: AgentRunStatus;
	prompt: string;
	max_runtime_seconds: number;
	push_branch: boolean;
	open_pull_request: boolean;
	result?: AgentRunResult;
	summary?: string;
	changed_files: string[];
	commits: string[];
	branch?: string;
	pull_request_url?: string;
	tests_run: string[];
	test_status: 'passed' | 'failed' | 'not_run';
	risk_notes: string[];
	exit_code?: number;
	error_message?: string;
	created_at: string;
	started_at?: string;
	completed_at?: string;
}

export interface AgentProvider {
	id: 'claude-code' | 'opencode' | 'codex' | 'custom';
	display_name: string;
	default_image: string;
	required_secrets: string[];
	supported_modes: Array<'interactive' | 'autonomous'>;
	custom: boolean;
}

export interface ResourceSample {
	id: number;
	cpu_percent: number;
	memory_bytes: number;
	disk_bytes: number;
	pids: number;
	network_rx_bytes: number;
	network_tx_bytes: number;
	created_at: string;
}

export interface CreateDevMachineInput {
	name?: string;
	issue_id?: string;
	project_id?: string;
	repo?: { provider: 'github'; owner: string; name: string; url: string };
	base_branch?: string;
	working_branch?: string;
	size: 'small' | 'medium' | 'large';
	services: { ide: boolean; browser: boolean };
	agents: Array<{ provider: AgentProvider['id']; mode: 'interactive' | 'autonomous'; config?: unknown }>;
	env_vars: Array<{
		name: string;
		value: string;
		target_service: 'ide' | 'agent' | 'collector';
		provider?: string;
		secret?: boolean;
	}>;
	environment_id?: string;
	keep_running: boolean;
	environment_builder?: boolean;
}

export interface CreateAgentRunInput {
	checkout_id?: string;
	use_root_workspace: boolean;
	provider: AgentProvider['id'];
	mode: 'interactive' | 'autonomous';
	prompt: string;
	acceptance_criteria: string[];
	allowed_commands: string[];
	forbidden_paths: string[];
	test_command: string[];
	max_runtime_seconds: number;
	allowed_secrets: string[];
	push_branch: boolean;
	open_pull_request: boolean;
}

export interface DevMachinePolicy {
	workspace_id: string;
	enabled: boolean;
	max_concurrent_machines: number;
	max_machines_per_user: number;
	max_daily_agent_runs: number;
	max_runtime_minutes: number;
	max_disk_gb: number;
	allowed_providers: string[];
	allowed_repositories: string[];
	allow_custom_providers: boolean;
	idle_pause_minutes: number;
}

export interface DevMachineCheckout {
	id: string;
	workspace_id: string;
	machine_id: string;
	issue_id: string;
	github_repo_id: string;
	repository_full_name: string;
	base_branch: string;
	working_branch: string;
	workspace_path: string;
	status: 'queued' | 'preparing' | 'ready' | 'failed';
	last_error?: string;
	created_at: string;
	updated_at: string;
	last_activity_at?: string;
}

export interface DevMachineEnvironment {
	id: string;
	workspace_id: string;
	name: string;
	image_ref: string;
	image_digest?: string;
	status: DevMachineEnvironmentStatus;
	source_machine_id?: string;
	created_by_user_id?: string;
	delete_requested_at?: string;
	created_at: string;
	updated_at: string;
}

export interface DevMachineScopeSetting {
	id?: string;
	workspace_id: string;
	team_id?: string;
	project_id?: string;
	issue_id?: string;
	github_repo_id?: string;
	base_branch?: string;
	environment_id?: string;
	created_at?: string;
	updated_at?: string;
}

export interface DevMachineTerminalSession {
	id: string;
	machine_id: string;
	checkout_id?: string;
	name: string;
	runtime_session_name: string;
	status: 'active' | 'closed' | string;
	created_at: string;
	last_activity_at: string;
	closed_at?: string;
}

export interface LaunchMachineServiceResponse {
	status: DevMachineLaunchStatus;
	launch_url?: string;
	expires_at?: string;
	operation?: DevMachineOperation;
	retry_after_seconds?: number;
}

export interface CreateTerminalSessionInput {
	name?: string;
	checkout_id?: string;
}

export interface TerminalSessionLaunchResponse {
	status: DevMachineLaunchStatus;
	session?: DevMachineTerminalSession;
	launch_url?: string;
	web_socket_url?: string;
	protocol?: 'ttyd.v1' | string;
	expires_at?: string;
	operation?: DevMachineOperation;
	retry_after_seconds?: number;
}

export interface BulkPermanentDeleteDevMachinesInput {
	machine_ids?: string[];
	older_than_days?: number;
	include_failed: boolean;
	include_expired: boolean;
}

export interface AgentRunStep {
	id: string;
	agent_run_id: string;
	sequence: number;
	step_type: string;
	name: string;
	status: 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled' | 'skipped';
	command_argv?: string[];
	summary?: string;
	exit_code?: number;
	started_at?: string;
	completed_at?: string;
	created_at: string;
}

export interface AgentRunTrace {
	run: AgentRun;
	steps: AgentRunStep[];
	events: DevMachineEvent[];
	logs: DevMachineLogChunk[];
	next_event_id: number;
	next_log_id: number;
	has_more_events: boolean;
	has_more_logs: boolean;
}

File: BE/internal/dto/dev_machine.go
package dto

import (
	"time"

	"github.com/kuayle/kuayle-backend/internal/domain"
)

type DevMachineRepoInput struct {
	Provider string `json:"provider" validate:"required,oneof=github"`
	Owner    string `json:"owner" validate:"required,max=255"`
	Name     string `json:"name" validate:"required,max=255"`
	URL      string `json:"url" validate:"required,url"`
}

type DevMachineServicesInput struct {
	IDE     bool `json:"ide"`
	Browser bool `json:"browser"`
}

type DevMachineAgentInput struct {
	Provider string `json:"provider" validate:"required,oneof=claude-code opencode codex custom"`
	Mode     string `json:"mode" validate:"required,oneof=interactive autonomous"`
	Config   any    `json:"config,omitempty"`
}

type DevMachineEnvVarInput struct {
	Name          string  `json:"name" validate:"required,max=255"`
	Value         string  `json:"value" validate:"required,max=65536"`
	TargetService string  `json:"target_service" validate:"required,oneof=ide agent collector"`
	Provider      *string `json:"provider,omitempty"`
	Secret        *bool   `json:"secret,omitempty"`
	TTLMinutes    *int    `json:"ttl_minutes,omitempty" validate:"omitempty,min=1,max=1440"`
}

type CreateDevMachineRequest struct {
	Name               string                  `json:"name" validate:"omitempty,min=3,max=255"`
	IssueID            *string                 `json:"issue_id,omitempty" validate:"omitempty,uuid"`
	ProjectID          *string                 `json:"project_id,omitempty" validate:"omitempty,uuid"`
	Repo               *DevMachineRepoInput    `json:"repo,omitempty"`
	BaseBranch         string                  `json:"base_branch,omitempty" validate:"omitempty,max=255"`
	WorkingBranch      string                  `json:"working_branch,omitempty" validate:"omitempty,max=255"`
	Size               string                  `json:"size" validate:"required,oneof=small medium large"`
	Services           DevMachineServicesInput `json:"services"`
	Agents             []DevMachineAgentInput  `json:"agents" validate:"max=4,dive"`
	EnvVars            []DevMachineEnvVarInput `json:"env_vars" validate:"max=64,dive"`
	EnvironmentID      *string                 `json:"environment_id,omitempty" validate:"omitempty,uuid"`
	KeepRunning        bool                    `json:"keep_running"`
	EnvironmentBuilder bool                    `json:"environment_builder"`
}

type DevMachineNameAvailabilityResponse struct {
	Name      string `json:"name"`
	Available bool   `json:"available"`
}

type DevMachineScopeSettingRequest struct {
	ScopeType     string  `json:"scope_type" validate:"required,oneof=workspace team project issue"`
	ScopeID       *string `json:"scope_id,omitempty" validate:"omitempty,uuid"`
	GitHubRepoID  *string `json:"github_repo_id,omitempty" validate:"omitempty,uuid"`
	BaseBranch    *string `json:"base_branch,omitempty" validate:"omitempty,max=255"`
	EnvironmentID *string `json:"environment_id,omitempty" validate:"omitempty,uuid"`
}

type CheckoutIssueRequest struct {
	IssueID string `json:"issue_id" validate:"required,uuid"`
}

type UpdateDevMachineRequest struct {
	KeepRunning *bool `json:"keep_running,omitempty"`
}

type BulkDeleteDevMachinesRequest struct {
	MachineIDs []string `json:"machine_ids,omitempty" validate:"omitempty,max=500,dive,uuid"`
}

type BulkDeleteDevMachineResult struct {
	MachineID string `json:"machine_id"`
	Status    string `json:"status"`
	ErrorCode string `json:"error_code,omitempty"`
}

type BulkDeleteDevMachinesResponse struct {
	Count     int                          `json:"count"`
	Requested int                          `json:"requested"`
	Results   []BulkDeleteDevMachineResult `json:"results"`
}

type PurgeDevMachinesRequest struct {
	MachineIDs     []string `json:"machine_ids,omitempty" validate:"omitempty,max=500,dive,uuid"`
	OlderThanDays  int      `json:"older_than_days,omitempty" validate:"omitempty,min=1,max=3650"`
	IncludeFailed  bool     `json:"include_failed"`
	IncludeExpired bool     `json:"include_expired"`
}

type CreateDevMachineEnvironmentRequest struct {
	Name            string `json:"name" validate:"required,min=3,max=255"`
	SourceMachineID string `json:"source_machine_id" validate:"required,uuid"`
}

type CreateTerminalSessionRequest struct {
	Name       string  `json:"name,omitempty" validate:"omitempty,max=128"`
	CheckoutID *string `json:"checkout_id,omitempty" validate:"omitempty,uuid"`
}

type TerminalSessionResponse struct {
	ID                 string     `json:"id"`
	MachineID          string     `json:"machine_id"`
	CheckoutID         *string    `json:"checkout_id,omitempty"`
	Name               string     `json:"name"`
	RuntimeSessionName string     `json:"runtime_session_name"`
	Status             string     `json:"status"`
	CreatedAt          time.Time  `json:"created_at"`
	LastActivityAt     time.Time  `json:"last_activity_at"`
	ClosedAt           *time.Time `json:"closed_at,omitempty"`
}

type TerminalSessionLaunchResponse struct {
	Status    string                   `json:"status"`
	Session   *TerminalSessionResponse `json:"session,omitempty"`
	LaunchURL string                   `json:"launch_url,omitempty"`
	// WebSocketURL is a one-use, frontend-origin-bound gateway URL for ttyd's
	// native /ws protocol. Native clients must speak ttyd framing directly: send
	// ttyd's initial terminal-size JSON message, then ttyd input/resize frames.
	WebSocketURL      string                       `json:"web_socket_url,omitempty"`
	Protocol          string                       `json:"protocol,omitempty"`
	ExpiresAt         time.Time                    `json:"expires_at,omitempty"`
	Operation         *DevMachineOperationResponse `json:"operation,omitempty"`
	RetryAfterSeconds int                          `json:"retry_after_seconds,omitempty"`
}

type DevMachineListParams struct {
	PaginationParams
	Status  string `query:"status"`
	IssueID string `query:"issue_id"`
}

type DevMachineOperationResponse struct {
	ID             string     `json:"id"`
	Action         string     `json:"action"`
	Status         string     `json:"status"`
	Generation     int64      `json:"generation"`
	IdempotencyKey string     `json:"idempotency_key"`
	Attempts       int        `json:"attempts"`
	ErrorCode      *string    `json:"error_code,omitempty"`
	ErrorMessage   *string    `json:"error_message,omitempty"`
	CreatedAt      time.Time  `json:"created_at"`
	CompletedAt    *time.Time `json:"completed_at,omitempty"`
}

type CreateAgentRunRequest struct {
	CheckoutID         *string  `json:"checkout_id,omitempty" validate:"omitempty,uuid"`
	UseRootWorkspace   bool     `json:"use_root_workspace"`
	Provider           string   `json:"provider" validate:"required,oneof=claude-code opencode codex custom"`
	Mode               string   `json:"mode" validate:"required,oneof=interactive autonomous"`
	Prompt             string   `json:"prompt" validate:"required,min=1,max=100000"`
	AcceptanceCriteria []string `json:"acceptance_criteria"`
	AllowedCommands    []string `json:"allowed_commands"`
	ForbiddenPaths     []string `json:"forbidden_paths"`
	TestCommand        []string `json:"test_command"`
	MaxRuntimeSeconds  int      `json:"max_runtime_seconds" validate:"omitempty,min=30,max=86400"`
	ExtraArgs          []string `json:"extra_args"`
	Config             any      `json:"config,omitempty"`
	AllowedSecrets     []string `json:"allowed_secrets"`
	PushBranch         *bool    `json:"push_branch,omitempty"`
	OpenPullRequest    bool     `json:"open_pull_request"`
}

type AgentProviderResponse struct {
	ID              string   `json:"id"`
	DisplayName     string   `json:"display_name"`
	DefaultImage    string   `json:"default_image"`
	RequiredSecrets []string `json:"required_secrets"`
	SupportedModes  []string `json:"supported_modes"`
	Custom          bool     `json:"custom"`
}

type LaunchServiceResponse struct {
	Status            string                       `json:"status"`
	LaunchURL         string                       `json:"launch_url,omitempty"`
	ExpiresAt         time.Time                    `json:"expires_at,omitempty"`
	Operation         *DevMachineOperationResponse `json:"operation,omitempty"`
	RetryAfterSeconds int                          `json:"retry_after_seconds,omitempty"`
}

type EventListParams struct {
	AfterID int64 `query:"after_id"`
	Limit   int   `query:"limit"`
}

func (p *EventListParams) Defaults() {
	if p.Limit <= 0 || p.Limit > 200 {
		p.Limit = 100
	}
}

type LogListParams struct {
	AfterID int64   `query:"after_id"`
	Limit   int     `query:"limit"`
	RunID   *string `query:"agent_run_id"`
}

func (p *LogListParams) Defaults() {
	if p.Limit <= 0 || p.Limit > 1000 {
		p.Limit = 250
	}
}

type DevMachinePolicyRequest struct {
	Enabled               bool     `json:"enabled"`
	MaxConcurrentMachines int      `json:"max_concurrent_machines" validate:"min=0,max=100"`
	MaxMachinesPerUser    int      `json:"max_machines_per_user" validate:"min=0,max=50"`
	MaxDailyAgentRuns     int      `json:"max_daily_agent_runs" validate:"min=0,max=10000"`
	MaxRuntimeMinutes     int      `json:"max_runtime_minutes" validate:"min=5,max=1440"`
	MaxDiskGB             int      `json:"max_disk_gb" validate:"min=20,max=2048"`
	AllowedProviders      []string `json:"allowed_providers"`
	AllowedRepositories   []string `json:"allowed_repositories"`
	AllowCustomProviders  bool     `json:"allow_custom_providers"`
	IdlePauseMinutes      int      `json:"idle_pause_minutes" validate:"omitempty,min=5,max=10080"`
}

type CollectorEventInput struct {
	AgentRunID *string    `json:"agent_run_id,omitempty" validate:"omitempty,uuid"`
	Source     string     `json:"source" validate:"required,oneof=agent browser collector filesystem git process shell"`
	EventType  string     `json:"event_type" validate:"required,max=128"`
	Payload    any        `json:"payload"`
	OccurredAt *time.Time `json:"occurred_at,omitempty"`
}

type CollectorLogInput struct {
	AgentRunID *string `json:"agent_run_id,omitempty" validate:"omitempty,uuid"`
	ServiceID  *string `json:"service_id,omitempty" validate:"omitempty,uuid"`
	Stream     string  `json:"stream" validate:"required,oneof=stdout stderr pty system"`
	Sequence   int64   `json:"sequence" validate:"min=0"`
	Content    string  `json:"content" validate:"required,max=65536"`
	Truncated  bool    `json:"truncated"`
}

type TraceListParams struct {
	EventsAfterID int64 `query:"events_after_id"`
	EventsLimit   int   `query:"events_limit"`
	LogsAfterID   int64 `query:"logs_after_id"`
	LogsLimit     int   `query:"logs_limit"`
}

func (p *TraceListParams) Defaults() {
	if p.EventsLimit <= 0 || p.EventsLimit > 500 {
		p.EventsLimit = 200
	}
	if p.LogsLimit <= 0 || p.LogsLimit > 2000 {
		p.LogsLimit = 500
	}
}

type AgentRunTraceResponse struct {
	Run           *domain.DevMachineAgentRun      `json:"run"`
	Steps         []domain.DevMachineAgentRunStep `json:"steps"`
	Events        []domain.DevMachineEvent        `json:"events"`
	Logs          []domain.DevMachineLogChunk     `json:"logs"`
	NextEventID   int64                           `json:"next_event_id"`
	NextLogID     int64                           `json:"next_log_id"`
	HasMoreEvents bool                            `json:"has_more_events"`
	HasMoreLogs   bool                            `json:"has_more_logs"`
}
